File: sdk/dart/dovahlink_client/lib/src/protocol/subscribe_payload.dart
import 'package:json_annotation/json_annotation.dart';

import 'package:dovahlink_client_sdk/src/protocol/json_map.dart';

part 'subscribe_payload.g.dart';

/// Outgoing `subscribe` payload (`protocol/schema/README.md`'s `subscribe`). [stateAreas] is the
/// client's complete desired active set; an empty list removes every state-area subscription.
/// Encode-only: the client never decodes its own `subscribe`.
@JsonSerializable(createFactory: false)
class SubscribePayload {
  /// The complete desired set of state areas for this connection.
  final List<String> stateAreas;

  /// Creates a subscribe payload.
  const SubscribePayload({required this.stateAreas});

  /// Encodes this payload as a JSON object.
  JsonMap toJson() => _$SubscribePayloadToJson(this);
}

File: integration/DovahLinkValidationClient/SubscribePayload.cs
using System.Text.Json.Nodes;

namespace DovahLinkValidationClient;

/// <summary>
/// Outgoing <c>subscribe</c> payload (<c>protocol/schema/README.md</c>'s <c>subscribe</c>).
/// Encode-only: this client never decodes its own <c>subscribe</c>.
/// </summary>
/// <param name="StateAreas">The state areas requested.</param>
public sealed record SubscribePayload(IReadOnlyList<string> StateAreas)
{
    /// <summary>
    /// Encodes this payload as a JSON object.
    /// </summary>
    /// <returns>The encoded <c>subscribe</c> payload.</returns>
    public JsonObject Encode()
    {
        var array = new JsonArray();
        foreach (string area in StateAreas)
        {
            array.Add(area);
        }
        return new JsonObject { ["stateAreas"] = array };
    }
}
